<h1 align=center><font size=5>Prediciting Engine Rating</font></h1>

<b>Objective:</b>
Task is to write a small Python or R script that predicts the engine rating based on the
inspection parameters using only the provided dataset. You need to find all the cases/outliers where
the rating has been given incorrectly as compared to current condition of the engine.
This task is designed to test your Python or R ability, your knowledge of Data Science techniques,
your ability to find trends, outliers, relative importance of variables with deviation in target variable
and your ability to work effectively, efficiently and independently within a commercial setting.
This task is designed as well to test your hyper-tuning abilities or lateral thinking.
 
Deliverables:
<ul>
    <li>         One Python or R script
    <li>         One requirements text file including an exhaustive list of packages and version numbers used
    in your solution
    <li>         Summary of your insights
    <li>         List of cases which are outliers/incorrectly rated as high or low and it should be backed with
    analysis/reasons.
    <li>        model object files for reproducibility.
</ul>


The solution should at a minimum do the following:
<ul>
    <li>         Load the data into memory
    <li>         Prepare the data for modelling
    <li>         EDA of the variables
    <li>         Build a model on training data
    <li>         Test the model on testing data
    <li>         Provide some measure of performance
    <li>         Outlier analysis and detection
</ul>

<h1 align=center><font size=5>Data Analysis with Python</font></h1>

# 1. Data Wrangling

<h2>What is the purpose of Data Wrangling?</h2>

Data Wrangling is the process of converting data from the initial format to a format that may be better for analysis.

<h3>Import data</h3>
<p>
The dataset for prediction of engine rating provided by Celebal Technologies.
</p>

<h4>Import pandas</h4> 

In [101]:
import pandas as pd
import matplotlib.pylab as plt

<h2>Reading the data set.</h2>

In [102]:
filename = 'data.xlsx' # Name of the file

In [103]:
excel_file = pd.ExcelFile(filename) # Provided excel file
fields_summary = pd.read_excel(excel_file, 'fields summary') # Sheet 1 with fields summary
data = pd.read_excel(excel_file, 'data') # Sheet 2 with the required dataset

The following is the list of independent variables with description:

In [104]:
fields_summary[:-1]

Columns                   Description  \
0                         appointmentId                     Unique id   
1                   inspectionStartTime               inspection date   
2                                  year             registration year   
3                                 month            registration month   
4      engineTransmission_battery_value  condition ok: yes\nelse : no   
..                                  ...                           ...   
67  engineTransmission_comments_value_2                      comments   
68  engineTransmission_comments_value_3                      comments   
69  engineTransmission_comments_value_4                      comments   
70                            fuel_type                     fuel type   
71                     odometer_reading                       mileage   

       Variable  
0   Independent  
1   Independent  
2   Independent  
3   Independent  
4   Independent  
..          ...  
67  Independent  
68  Independent  
69  Independent  
70  Independent  
71  Independent  

[72 rows x 3 columns]

This is the dependent variable with description:

In [105]:
fields_summary[-1:]

Columns    Description   Variable
72  rating_engineTransmission  engine rating  Dependent

As the names of the columns seem convenient we don't have a requirement to rename the header's.

In [106]:
# To see what the data set looks like, we'll use the head() method.
data.head()

appointmentId inspectionStartTime  year  month  \
0         aj_01 2019-02-03 15:43:48  2008      8   
1         aj_02 2019-01-16 13:02:12  2007      5   
2         aj_03 2019-02-09 13:31:04  2012      5   
3         aj_04 2019-01-18 11:02:53  2013      1   
4         aj_05 2019-01-27 12:12:06  2011      7   

  engineTransmission_battery_value engineTransmission_battery_cc_value_0  \
0                               No                                  Weak   
1                              Yes                                   NaN   
2                              Yes                                   NaN   
3                              Yes                                   NaN   
4                              Yes                                   NaN   

  engineTransmission_battery_cc_value_1 engineTransmission_battery_cc_value_2  \
0                                   NaN                                   NaN   
1                                   NaN                                   NaN   
2                                   NaN                                   NaN   
3                                   NaN                                   NaN   
4                                   NaN                                   NaN   

  engineTransmission_battery_cc_value_3 engineTransmission_battery_cc_value_4  \
0                                   NaN                                   NaN   
1                                   NaN                                   NaN   
2                                   NaN                                   NaN   
3                                   NaN                                   NaN   
4                                   NaN                                   NaN   

   ... engineTransmission_gearShifting_cc_value_1  \
0  ...                                        NaN   
1  ...                                        NaN   
2  ...                                        NaN   
3  ...                                        NaN   
4  ...                                        NaN   

  engineTransmission_gearShifting_cc_value_2  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  engineTransmission_comments_value_0  \
0                          Pickup Low   
1                                 NaN   
2                                 NaN   
3                                 NaN   
4                                 NaN   

       engineTransmission_comments_value_1  \
0  Starter Motor / Solenoid malfunctioning   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

  engineTransmission_comments_value_2 engineTransmission_comments_value_3  \
0                                 NaN                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

  engineTransmission_comments_value_4     fuel_type odometer_reading  \
0                                 NaN  Petrol + CNG           124600   
1                                 NaN  Petrol + CNG            78108   
2                                 NaN        Diesel            93348   
3                                 NaN  Petrol + CNG            48233   
4                                 NaN  Petrol + CNG            74368   

  rating_engineTransmission  
0                       3.0  
1                       4.0  
2                       3.5  
3                       3.5  
4                       4.0  

[5 rows x 73 columns]

As we can see, several NaN values appeared in the dataframe; those are missing values which may hinder our further analysis. 
<div>So, how do we identify all those missing values and deal with them?</div> 


<b>How to work with missing data?</b>

Steps for working with missing data:
<ol>
    <li>Identify missing data</li>
    <li>Deal with missing data</li>
    <li>Correct data format</li>
</ol>

<h2 id="identify_handle_missing_values">Identify and handle missing values</h2>


<h3 id="identify_missing_values">Identify missing values</h3>
<h4>Convert any irrevlevant values to NaN</h4>
In any dataset, mostly missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Here we would have used the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B.

But as we see that there are no question marks or irrelant values in the dataset and just NaN values, hence we proceed to the next step.

<h4>Evaluating for Missing Data</h4>

The missing values are converted to Python's default. We use Python's built-in functions to identify these missing values. There are two methods to detect missing data:
<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [107]:
missing_data = data.isnull()
missing_data.head(5)

appointmentId  inspectionStartTime   year  month  \
0          False                False  False  False   
1          False                False  False  False   
2          False                False  False  False   
3          False                False  False  False   
4          False                False  False  False   

   engineTransmission_battery_value  engineTransmission_battery_cc_value_0  \
0                             False                                  False   
1                             False                                   True   
2                             False                                   True   
3                             False                                   True   
4                             False                                   True   

   engineTransmission_battery_cc_value_1  \
0                                   True   
1                                   True   
2                                   True   
3                                   True   
4                                   True   

   engineTransmission_battery_cc_value_2  \
0                                   True   
1                                   True   
2                                   True   
3                                   True   
4                                   True   

   engineTransmission_battery_cc_value_3  \
0                                   True   
1                                   True   
2                                   True   
3                                   True   
4                                   True   

   engineTransmission_battery_cc_value_4  ...  \
0                                   True  ...   
1                                   True  ...   
2                                   True  ...   
3                                   True  ...   
4                                   True  ...   

   engineTransmission_gearShifting_cc_value_1  \
0                                        True   
1                                        True   
2                                        True   
3                                        True   
4                                        True   

   engineTransmission_gearShifting_cc_value_2  \
0                                        True   
1                                        True   
2                                        True   
3                                        True   
4                                        True   

   engineTransmission_comments_value_0  engineTransmission_comments_value_1  \
0                                False                                False   
1                                 True                                 True   
2                                 True                                 True   
3                                 True                                 True   
4                                 True                                 True   

   engineTransmission_comments_value_2  engineTransmission_comments_value_3  \
0                                 True                                 True   
1                                 True                                 True   
2                                 True                                 True   
3                                 True                                 True   
4                                 True                                 True   

   engineTransmission_comments_value_4  fuel_type  odometer_reading  \
0                                 True      False             False   
1                                 True      False             False   
2                                 True      False             False   
3                                 True      False             False   
4                                 True      False             False   

   rating_engineTransmission  
0                      False  
1                      False  
2                      False  
3                     

"True" stands for missing value, while "False" stands for not missing value.

<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values. 
</p>

In [108]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

appointmentId
False    26307
Name: appointmentId, dtype: int64

inspectionStartTime
False    26307
Name: inspectionStartTime, dtype: int64

year
False    26307
Name: year, dtype: int64

month
False    26307
Name: month, dtype: int64

engineTransmission_battery_value
False    26307
Name: engineTransmission_battery_value, dtype: int64

engineTransmission_battery_cc_value_0
True     22869
False     3438
Name: engineTransmission_battery_cc_value_0, dtype: int64

engineTransmission_battery_cc_value_1
True     25877
False      430
Name: engineTransmission_battery_cc_value_1, dtype: int64

engineTransmission_battery_cc_value_2
True     26235
False       72
Name: engineTransmission_battery_cc_value_2, dtype: int64

engineTransmission_battery_cc_value_3
True     26291
False       16
Name: engineTransmission_battery_cc_value_3, dtype: int64

engineTransmission_battery_cc_value_4
True     26303
False        4
Name: engineTransmission_battery_cc_value_4, dtype: int64

engineTransmission_engineoilL

Based on the summary above, each column has 26307 rows of data, fifty four columns containing missing data:
<ol>
    <li>"engineTransmission_battery_cc_value_0": 22869 missing data</li>
    <li>"engineTransmission_battery_cc_value_1": 25877 missing data</li>
    <li>"engineTransmission_battery_cc_value_2": 26235 missing data</li>
    <li>"engineTransmission_battery_cc_value_3": 26291 missing data</li>
    <li>"engineTransmission_battery_cc_value_4": 26303 missing data</li>
    <li>"engineTransmission_engineOilLevelDipstick_cc_value_0": 25896 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_0": 7750 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_1": 15303 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_2": 19714 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_3": 22565 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_4": 24535 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_5": 25698 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_6": 26186 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_7": 26296 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_8": 26305 missing data</li>
    <li>"engineTransmission_engineOil_cc_value_9": 26307 missing data</li>
    <li>"engineTransmission_engine_cc_value_0": 17237 missing data</li>
    <li>"engineTransmission_engine_cc_value_1": 21223 missing data</li>
    <li>"engineTransmission_engine_cc_value_2": 23933 missing data</li>
    <li>"engineTransmission_engine_cc_value_3": 25403 missing data</li>
    <li>"engineTransmission_engine_cc_value_4": 26011 missing data</li>
    <li>"engineTransmission_engine_cc_value_5": 26215 missing data</li>
    <li>"engineTransmission_engine_cc_value_6": 26270 missing data</li>
    <li>"engineTransmission_engine_cc_value_7": 26299 missing data</li>
    <li>"engineTransmission_engine_cc_value_8": 26303 missing data</li>
    <li>"engineTransmission_engine_cc_value_9": 26304 missing data</li>
    <li>"engineTransmission_engine_cc_value_10": 26307 missing data</li>
    <li>"engineTransmission_coolant_cc_value_0": 16271 missing data</li>
    <li>"engineTransmission_coolant_cc_value_1": 23944 missing data</li>
    <li>"engineTransmission_coolant_cc_value_2": 26072 missing data</li>
    <li>"engineTransmission_coolant_cc_value_3": 26299 missing data</li>
    <li>"engineTransmission_engineMounting_cc_value_0": 17020 missing data</li>
    <li>"engineTransmission_engineSound_cc_value_0": 7227 missing data</li>
    <li>"engineTransmission_engineSound_cc_value_1": 12802 missing data</li>
    <li>"engineTransmission_engineSound_cc_value_2": 20235 missing data</li>
    <li>"engineTransmission_engineSound_cc_value_3": 24155 missing data</li>
    <li>"engineTransmission_engineSound_cc_value_4": 25580 missing data</li>
    <li>"engineTransmission_engineSound_cc_value_5": 26123 missing data</li>
    <li>"engineTransmission_exhaustSmoke_cc_value_0": 20751 missing data</li>
    <li>"engineTransmission_clutch_cc_value_0": 14543 missing data</li>
    <li>"engineTransmission_clutch_cc_value_1": 22003 missing data</li>
    <li>"engineTransmission_clutch_cc_value_2": 25510 missing data</li>
    <li>"engineTransmission_clutch_cc_value_3": 26148 missing data</li>
    <li>"engineTransmission_clutch_cc_value_4": 26264 missing data</li>
    <li>"engineTransmission_clutch_cc_value_5": 26300 missing data</li>
    <li>"engineTransmission_clutch_cc_value_6": 26306 missing data</li>
    <li>"engineTransmission_gearShifting_cc_value_0": 22843 missing data</li>
    <li>"engineTransmission_gearShifting_cc_value_1": 25536 missing data</li>
    <li>"engineTransmission_gearShifting_cc_value_2": 26241 missing data</li>
    <li>"engineTransmission_comments_value_0": 22241 missing data</li>
    <li>"engineTransmission_comments_value_1": 24999 missing data</li>
    <li>"engineTransmission_comments_value_2": 25917 missing data</li>
    <li>"engineTransmission_comments_value_3": 26248 missing data</li>
    <li>"engineTransmission_comments_value_4": 26293 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
        c. replace it based on other functions
    </li>
</ol>

Whole columns should be dropped only if most entries in the column are empty. In our dataset, moat of the columns are empty enough to drop entirely.
Just because here we have most of the columns containing missing values, we need to provide a threshold for the applicability of column in building our model.
As we know all the columns above which have missing values are represented with the first column without the numerical postfix name as "Yes" if applicable and "No" if there is any kind of a problem. Therefore, we can safely drop all the columns above.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

## Drop missing values

In [109]:
data.drop(columns = ['engineTransmission_battery_cc_value_0','engineTransmission_battery_cc_value_1','engineTransmission_battery_cc_value_2','engineTransmission_battery_cc_value_3','engineTransmission_battery_cc_value_4','engineTransmission_engineOilLevelDipstick_cc_value_0','engineTransmission_engineOil_cc_value_0','engineTransmission_engineOil_cc_value_1','engineTransmission_engineOil_cc_value_2','engineTransmission_engineOil_cc_value_3','engineTransmission_engineOil_cc_value_4','engineTransmission_engineOil_cc_value_5','engineTransmission_engineOil_cc_value_6','engineTransmission_engineOil_cc_value_7','engineTransmission_engineOil_cc_value_8','engineTransmission_engineOil_cc_value_9','engineTransmission_engine_cc_value_0','engineTransmission_engine_cc_value_1','engineTransmission_engine_cc_value_2','engineTransmission_engine_cc_value_3','engineTransmission_engine_cc_value_4','engineTransmission_engine_cc_value_5','engineTransmission_engine_cc_value_6','engineTransmission_engine_cc_value_7','engineTransmission_engine_cc_value_8','engineTransmission_engine_cc_value_9','engineTransmission_engine_cc_value_10','engineTransmission_coolant_cc_value_0','engineTransmission_coolant_cc_value_1','engineTransmission_coolant_cc_value_2','engineTransmission_coolant_cc_value_3','engineTransmission_engineMounting_cc_value_0','engineTransmission_engineSound_cc_value_0','engineTransmission_engineSound_cc_value_1','engineTransmission_engineSound_cc_value_2','engineTransmission_engineSound_cc_value_3','engineTransmission_engineSound_cc_value_4','engineTransmission_engineSound_cc_value_5','engineTransmission_exhaustSmoke_cc_value_0','engineTransmission_clutch_cc_value_0','engineTransmission_clutch_cc_value_1','engineTransmission_clutch_cc_value_2','engineTransmission_clutch_cc_value_3','engineTransmission_clutch_cc_value_4','engineTransmission_clutch_cc_value_5','engineTransmission_clutch_cc_value_6','engineTransmission_gearShifting_cc_value_0','engineTransmission_gearShifting_cc_value_1','engineTransmission_gearShifting_cc_value_2','engineTransmission_comments_value_0','engineTransmission_comments_value_1','engineTransmission_comments_value_2','engineTransmission_comments_value_3','engineTransmission_comments_value_4','engineTransmission_engineBlowByBackCompression_cc_value_0'], inplace = True)

## Check dataframe after dropping the columns.

In [111]:
data.head()

appointmentId inspectionStartTime  year  month  \
0         aj_01 2019-02-03 15:43:48  2008      8   
1         aj_02 2019-01-16 13:02:12  2007      5   
2         aj_03 2019-02-09 13:31:04  2012      5   
3         aj_04 2019-01-18 11:02:53  2013      1   
4         aj_05 2019-01-27 12:12:06  2011      7   

  engineTransmission_battery_value  \
0                               No   
1                              Yes   
2                              Yes   
3                              Yes   
4                              Yes   

  engineTransmission_engineoilLevelDipstick_value  \
0                                             Yes   
1                                             Yes   
2                                             Yes   
3                                             Yes   
4                                             Yes   

  engineTransmission_engineOil engineTransmission_engine_value  \
0                           No                              No   
1                           No                              No   
2                           No                             Yes   
3                           No                             Yes   
4                           No                              No   

  engineTransmission_coolant_value engineTransmission_engineMounting_value  \
0                               No                                      No   
1                               No                                     Yes   
2                               No                                      No   
3                               No                                     Yes   
4                              Yes                                      No   

  engineTransmission_engineSound_value engineTransmission_exhaustSmoke_value  \
0                                   No                                    No   
1                                   No                                    No   
2                                   No                                    No   
3                                  Yes                                   Yes   
4                                   No                                    No   

  engineTransmission_engineBlowByBackCompression_value  \
0                                                 No     
1                                                Yes     
2                                                Yes     
3                                                Yes     
4                                                Yes     

  engineTransmission_clutch_value engineTransmission_gearShifting_value  \
0                              No                                   Yes   
1                             Yes                                   Yes   
2                              No                                   Yes   
3                              No                                   Yes   
4                              No                                   Yes   

      fuel_type  odometer_reading  rating_engineTransmission  
0  Petrol + CNG            124600                        3.0  
1  Petrol + CNG             78108                        4.0  
2        Diesel             93348                        3.5  
3  Petrol + CNG             48233                        3.5  
4  Petrol + CNG             74368                        4.0

Now we have no rows or columns with NaN values.

<h3 id="correct_data_format">Correct data format</h3>
<b>We are almost there!</b>
<p>The last step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other).</p>

In Pandas, we use 
<p><b>.dtype()</b> to check the data type</p>
<p><b>.astype()</b> to change the data type</p>

<h4>Lets list the data types for each column</h4>

In [113]:
data.dtypes

appointmentId                                                   object
inspectionStartTime                                     datetime64[ns]
year                                                             int64
month                                                            int64
engineTransmission_battery_value                                object
engineTransmission_engineoilLevelDipstick_value                 object
engineTransmission_engineOil                                    object
engineTransmission_engine_value                                 object
engineTransmission_coolant_value                                object
engineTransmission_engineMounting_value                         object
engineTransmission_engineSound_value                            object
engineTransmission_exhaustSmoke_value                           object
engineTransmission_engineBlowByBackCompression_value            object
engineTransmission_clutch_value                                 object
engine

<p>As we can see above, some columns are not of the correct data type. Numerical variables should have type 'float' or 'int', and variables with strings such as categories should have type 'object'. For example, 'engineTransmission_battery_value' and 'engineTransmission_engineoilLevelDipstick_value' variables are categorical values that describe the engines, so we should expect them to be of the type 'integers' or we can say dummy values; however, they are shown as type 'object'. We have to convert data types into a proper format for each column using the "astype()" method.</p> 

<h4>Convert data types to proper format by replacing "Yes" with 1 and "No" with 0.</h4>

In [121]:
data.replace("Yes",1, inplace = True) # Replacing "Yes" with 1.
data.replace("No",0, inplace = True) # Replacing "No" with 0.
data.head() # Checking dataframe after changing values.

appointmentId inspectionStartTime  year  month  \
0         aj_01 2019-02-03 15:43:48  2008      8   
1         aj_02 2019-01-16 13:02:12  2007      5   
2         aj_03 2019-02-09 13:31:04  2012      5   
3         aj_04 2019-01-18 11:02:53  2013      1   
4         aj_05 2019-01-27 12:12:06  2011      7   

   engineTransmission_battery_value  \
0                                 0   
1                                 1   
2                                 1   
3                                 1   
4                                 1   

   engineTransmission_engineoilLevelDipstick_value  \
0                                                1   
1                                                1   
2                                                1   
3                                                1   
4                                                1   

   engineTransmission_engineOil  engineTransmission_engine_value  \
0                             0                                0   
1                             0                                0   
2                             0                                1   
3                             0                                1   
4                             0                                0   

   engineTransmission_coolant_value  engineTransmission_engineMounting_value  \
0                                 0                                        0   
1                                 0                                        1   
2                                 0                                        0   
3                                 0                                        1   
4                                 1                                        0   

   engineTransmission_engineSound_value  \
0                                     0   
1                                     0   
2                                     0   
3                                     1   
4                                     0   

   engineTransmission_exhaustSmoke_value  \
0                                      0   
1                                      0   
2                                      0   
3                                      1   
4                                      0   

   engineTransmission_engineBlowByBackCompression_value  \
0                                                  0      
1                                                  1      
2                                                  1      
3                                                  1      
4                                                  1      

   engineTransmission_clutch_value  engineTransmission_gearShifting_value  \
0                                0                                      1   
1                                1                                      1   
2                                0                                      1   
3                                0                                      1   
4                                0                                      1   

      fuel_type  odometer_reading  rating_engineTransmission  
0  Petrol + CNG            124600                        3.0  
1  Petrol + CNG             78108                        4.0  
2        Diesel             93348                        3.5  
3  Petrol + CNG             48233                        3.5  
4  Petrol + CNG             74368                        4.0

## Checking datatypes after replacement of the values.

In [118]:
data.dtypes

appointmentId                                                   object
inspectionStartTime                                     datetime64[ns]
year                                                             int64
month                                                            int64
engineTransmission_battery_value                                 int64
engineTransmission_engineoilLevelDipstick_value                  int64
engineTransmission_engineOil                                     int64
engineTransmission_engine_value                                  int64
engineTransmission_coolant_value                                 int64
engineTransmission_engineMounting_value                          int64
engineTransmission_engineSound_value                             int64
engineTransmission_exhaustSmoke_value                            int64
engineTransmission_engineBlowByBackCompression_value             int64
engineTransmission_clutch_value                                  int64
engine

Now, we finally obtain the cleaned dataset with no missing values and all data in its proper format.

<h2 id="data_standardization">Data Standardization</h2>
<p>
Data is usually collected from different agencies with different formats.
(Data Standardization is also a term for a particular type of data normalization, where we subtract the mean and divide by the standard deviation)
</p>
    
<b>What is Standardization?</b>
<p>Standardization is the process of transforming data into a common format which allows the researcher to make the meaningful comparison. But, in our dataset as most of the value are categorical and the remaining are already standardized, we can move on to the next step.
</p>


<h2 id="data_normalization">Data Normalization</h2>

<b>Why normalization?</b>
<p>Normalization is the process of transforming values of several variables into a similar range. Typical normalizations include scaling the variable so the variable average is 0, scaling the variable so the variance is 1, or scaling variable so the variable values range from 0 to 1
</p>

<b>Example</b>
<p>To demonstrate normalization, let's say we want to scale some columns "x", "y" and "z" </p>
<p><b>Target:</b> We would like to Normalize those variables so their value ranges from 0 to 1.</p>
<p><b>Approach:</b> replace original value by (original value)/(maximum value)</p>

Here we have one column "odometer_reading" which can be normalized because of the presence of large values. Therefore, let's normalize the only possible column in our dataset.

In [123]:
# replace (original value) by (original value)/(maximum value)
data['odometer_reading'] = data['odometer_reading']/data['odometer_reading'].max()

## Save the cleaned dataset as clean_df.

In [138]:
data.to_csv('clean_df.csv', index = False)

Now, we have completed the data wrangling. So, let's start with Exploratory Data Analysis.